extensions/amp-truncate-text/0.1/amp-truncate-text.js

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-truncate-text-0.1.css';
+import {
+  CSS as ShadowCSS,
+} from '../../../build/amp-truncate-text-shadow-0.1.css';
+import {dev} from '../../../src/log';
+import {htmlFor} from '../../../src/static-template';
+import {isExperimentOn} from '../../../src/experiments';
+import {iterateCursor} from '../../../src/dom';
+import {truncateText} from './truncate-text';
+
+/**
+ * TODO(sparhami) List of stuff to do / consider:
+ * - Delay truncateing for things outside of the viewport
+ * - Only truncate a few things in a single pass, and defer others
+ * - If estimation + mutation takes too long, fall back to gradient
+ *   or perhaps nothing and position absolute the button on top of
+ *   text
+ *     * Maybe let the developer specify the gradient
+ * - If we had some rough bucket of performance, maybe just fallback
+ *   immediately to gradient / hard cut off.
+ * - Custom fonts can cause truncation to end up being wrong
+ *   when they load
+ *     * Can we just wait to layout if we know a font is loading?
+ *       Since all fonts are statically declared in AMP, this is just a
+ *       one time thing
+ */
+export class AmpTruncateText extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.content_ = null;
+
+    /** @private {?Element} */
+    this.expandSlot_ = null;
+
+    /** @private {?Element} */
+    this.collapseSlot_ = null;
+
+    /** @private {boolean} */
+    this.useShadow_ = false;
+
+    /** @private {?MutationObserver} */
+    this.mutationObserver_ = null;
+
+    /** @private {boolean} */
+    this.truncateRequested_ = false;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.useShadow_ = !!this.element.attachShadow &&
+      isExperimentOn(this.win, 'amp-truncate-text-shadow');
+
+    if (this.useShadow_) {
+      this.buildShadow_();
+    } else {
+      this.build_();
+    }
+
+    if ('MutationObserver' in window) {
```

Comment:
Is this ever not true in our environments?

Reply:
I'd rather we removed this. At this stage, I'd rather get an exception telling me where we may have a problem then just skipping mutation observer code entirely.

Reply:
Agreed. Removed the check.